server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    google.protobuf.Duration wait_timeout = 7;
+}
+
+message UpdateWorkflowResponse {
+    UpdateResultType result_type = 1;
+    string update_id = 2;
+    temporal.api.common.v1.Payloads output = 3;
+}
+
+enum UpdateResultType {
+    UPDATE_RESULT_TYPE_UNSPECIFIED = 0;
+    UPDATE_RESULT_TYPE_SUCCESS = 1;
+    UPDATE_RESULT_TYPE_ERROR = 2;
+}
+
+message AwaitUpdateResponseRequest {
+    string update_id = 1; 
+    google.protobuf.Duration wait_timeout = 2;
+}
+```
+
+## Revisiting Use Cases With Propsed API
+
+### Short Update
+
+Client calls `UpdateWorkflowRequest` with a suitable `wait_timeout`.
+
+### Uncertain Duration Update
+
+Client calls `UpdateWorkflowRequest` with an acceptable `wait_timeout`. If the
+`result_type` of the returned `UpdateWorkflowResponse` is
+`UPDATE_RESULT_TYPE_UNSPECIFIED` then the client can infer that the Update is
+still runing and can use the returned `update_id` to obtain the UpdateResponse
+at a later time with `AwaitUpdateWorkflowResponse`. If the `result_type` is any
+other value then the response `output` can be used directly.
+
+### Known Long Update
+
+Client calls `UpdateWorkflowRequest` with a zero `wait_timeout` and uses the
+returned `update_id` to obtain the UpdateResponse at a later time with
+`AwaitWorkflowResponse`.
+
+### Invalid Update
+
+If the UpdateRequest is rejected by the workflow execution while the RPC client
+is still connected, a suitable gRPC error can be returned synchronously. Note
+however rejected UpdateRequests are _not_ recorded in workflow history.
+
+### "Failing" Update
+
+An Update that is passes verification by the workflow code and runs to
+completion but encounters an error at the application level is still considered
+to have completed correctly from the perspective of the Temporal system. The
+application-level error is recorded as the `UpdateWorkflowResponse.output` and
+the `UpdateWorkflowResponse.result_type` is set to `UPDATE_RESULT_TYPE_ERROR`.
+
+### Unexpected Disconnect
+
+When a client is waiting on an UpdateResponse having invoked the
+`UpdateWorkflow` and it loses connectivity prior to receiving an `update_id`, it
+can reissue the same RPC request with the exact same `request_id`. The second
+request will either be new to the server (if the first request never made it
+there) or will be de-duplicated on the server-side such that the client gets
+attached to the execution of the previously submitted request. In either case,
+the `UpdateWorkflow` RPC call proceeds and returns normally from the client's
+perspective.
+
+
+## Concurrent Semantics
+
+Update processing both within a single workflow execution and across multiple
+workflow executions can procede in parallel however at most one update will be
+processed by any single workflow execution at a time. Updates that overlap in
```

Comment:
Does this mean my update handler may not be async in the SDK? Meaning, can I get an update request, execute an activity, and use the result of that activity as the update response?

Reply:
IIUC, yes, that's the intent. Seems like good flexibility to have, no? Or have I missed your point?

Reply:
Can you clarify your response - can update handlers be async?

Assuming you meant that update handlers _can_ be async for purposes of flexibility, that makes sense, I was just making sure that "at most one update will be processed by any single workflow execution at a time" didn't mean "processed through to update response" but instead means "processed but the update result can be provided later".

Reply:
Since the issue you are working on here is SO simple and clearly not challenging enough ;-) I just wanted to ask: are y'all considering the tangentially related question of a streaming model where the SDK talks to the backend over a durable connection (in some cases)?  There are lots of use cases where a single Workflow goes through a (potentially quite large) flurry of transitions and we would like to be able to process these in a (dramatically!) more-efficient fashion.  Obviously we're talking about a rather large change here but the payoff is huge (in some cases) so we have to at least discuss it.  Not sure now is the time, but the work you're doing right now brought it to mind ... thought you'd want to know it's on the table.

Reply:
@cretz My intent was to say that yes, updated handlers can be async

@paulnpdev Not sure exactly what you're getting at with "durable connection" but most gRPC clients multiplex lots of messages onto "transport" instances (roughly: connections) which have a longer lifecycle.

Reply:
@paulnpdev We do consider pipelining all the way to the workflow, but it is out of scope for this particular proposal.

Reply:
@mfateev Roger that.  FYI, I wasn't trying to co-opt this proposal to include the pipelining feature, just to bring it to the attention of the people involved.  It's entirely possible that the code can be written in a way that leaves the door (more?) open to that if the author is aware of the potential need.  I know that if I was working on this layer I would rather know than not.

Reply:
> Not sure exactly what you're getting at with "durable connection" but most gRPC clients multiplex lots of messages onto "transport" instances (roughly: connections) which have a longer lifetime than the request-response of a single RPC.

@mmcshane  As max points out we're not considering pipelining "right away" but just for the record, I'm aware of the underlying mechanism you describe.  The overhead I want to eliminate (someday:) is the VERY expensive stack on top of the underlying TCP connection.

Reply:
Yeah, that's a different discussion. Not sure which aspects of the stack above TCP are of greatest concern to you but personally I'd include TCP itself in the assessment.